pkg/controller/statusmanager/pod_status.go

```diff
@@ -0,0 +1,278 @@
+package statusmanager
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"os"
+	"reflect"
+	"strings"
+	"time"
+
+	configv1 "github.com/openshift/api/config/v1"
+
+	appsv1 "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/retry"
+	client "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	// if a rollout has not made any progress by this time,
+	// mark ourselves as Degraded
+	ProgressTimeout = 10 * time.Minute
+
+	// lastSeenAnnotation - the annotation where we stash our state
+	lastSeenAnnotation = "network.operator.openshift.io/last-seen-state"
+)
+
+// podState is a snapshot of the last-seen-state and last-changed-times
+// for pod-creating entities
+type podState struct {
+	// "public" for marshalling to json, since we can't have complex keys
+	DaemonsetStates  []daemonsetState
+	DeploymentStates []deploymentState
+
+	// maps for internal use
+	daemonsetState  map[types.NamespacedName]daemonsetState
+	deploymentState map[types.NamespacedName]deploymentState
+}
+
+// daemonsetState is the internal state we use to check if a rollout has
+// stalled. Since we don't get a last-changed time from the apiserver, we
+// need to synthesize it ourselves. This means that we may flap between
+// Degraded if the operator restarts.
+type daemonsetState struct {
+	LastSeenStatus appsv1.DaemonSetStatus
+	LastChangeTime time.Time
+	Name           types.NamespacedName
```

Comment:
maybe inline the `types.NamespacedName` field?

Reply:
Sure